modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/tree/BPlusTree.java

```diff
@@ -4802,14 +4802,21 @@ private boolean releaseForRetry(Tail<L> t) {
 
                     assert needReplaceInner != TRUE;
 
-                    if (tail.getCount() == 0 && tail.lvl != 0 && getRootLevel() == tail.lvl) {
+                    // Loop is needed to prevent the rare case when, after parallel remove of keys, empty root remains.
+                    // B+tree after removes key: [empty_root] - [empty_inner_node] - [5] ==>
+                    // B+tree after cutting empty root: [5]
+                    for (Tail<L> t = tail; t != null && t.getCount() == 0 && t.lvl != 0 && getRootLevel() == t.lvl; ) {
                         // Free root if it became empty after merge.
-                        cutRoot(tail.lvl);
-                        freePage(tail.pageId, tail.page, tail.buf, tail.walPlc, false);
+
+                        cutRoot(t.lvl);
+                        freePage(t.pageId, t.page, t.buf, t.walPlc, false);
+
+                        t = t.down;
 
                         // Exit: we are done.
                     }
-                    else if (tail.sibling != null &&
+
+                    if (tail.sibling != null &&
```

Comment:
`tail` can point on removed page here, isn't it? I bet there's a bug here. Why do you need additional variable `t` in the loop above? You could just use `tail` and a `while` loop instead, this way `tail` value would be correct. Maybe I'm wrong here, please correct me if that's the case.

Reply:
Correct it